merkledag/utils/utils.go

```diff
@@ -0,0 +1,113 @@
+package dagutils
+
+import (
+	"errors"
+	"time"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+func AddLink(ctx context.Context, ds dag.DAGService, root *dag.Node, childname string, childk key.Key) (*dag.Node, error) {
+	if childname == "" {
+		return nil, errors.New("cannot create link with no name!")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
```

Comment:
no timeout here. caller's responsibility. 

if you're worried about the "running out of time" problem (e.g. user gives context with timeout T, to do K steps, but we can estimate that if step 1 takes  > T/K time, we should fail early and avoid wasting resources on a doomed-to-run-out-of-time operation) use https://godoc.org/github.com/jbenet/go-context/frac